### PR TITLE
Feature/user roles

### PR DESCRIPTION
**Pull request by axeloz**
This feature add an authorization layer on authentication. There are two levels: 

* globally where a user can be a super administrator of the application. This admin can access the admin panel, edit any setting, add projects, groups and users and view and edit any projet.
* locally per project where a user can be either "regular user" or "manager". A manager can view, edit  and deploy the projects he has access to when a regular user can only view and deploy the projects he has access to. None of them can access the admin panel

Changes are the following:
* users listed into the admin panel's users page can be set as "administrators".
* projects listed into the admin panel's projects page can have zero, one or more managers and/or regular users attached to them.
* projects pages are now protected by Laravel authorizations: a non-admin user can only access the projects he has authorization for. 
* projects views local features are now protected by the `@can` Laravel's feature to allow edition of some settings.
* installation process now set the first user as admin
* the `deployer:create-user` command now accepts the `--admin` option in order to define a new user as admin
* the Laravel's seeders now include a user with the admin role.

Todos: 
* the autocomplete of users (using `bootstrap-tagsinput` and `bootstrap-3-typeahead`) to be attached to a project does work but the data source IS NOT the real actual users list. The users collection has been added into the view but I didn't get how to use this collection from the Projects view.
* test the deployment process on behalf of a "regular user" and "manager" user 
* prevent an administrator to attach the same user as "regular user" and "manager" for the same project

### CODE REVIEW

File: app/Console/Commands/CreateUser.php [hunk 1]
```diff
@@ -77,6 +78,8 @@ public function handle(Dispatcher $dispatcher, Validation $validation)
 
         $send_email = (!$this->option('no-email'));
 
+        $arguments['is_admin']   = (!$this->option('admin') ? 0 : 1);
```
Comment (REBELinBLUE) [inline app/Console/Commands/CreateUser.php]: Could you change this to `$arguments['is_admin']   = (!$this->option('admin'));` as with the previous line

File: app/Console/Commands/InstallApp.php [hunk 2]
```diff
@@ -360,7 +360,7 @@ private function generateKey()
      */
     private function createAdminUser($name, $email, $password)
     {
-        $process = $this->artisanProcess('deployer:create-user', [$name, $email, $password, '--no-email']);
+        $process = $this->artisanProcess('deployer:create-user', [$name, $email, $password, '--no-email --admin']);
```
Comment (REBELinBLUE) [inline app/Console/Commands/InstallApp.php]: Can you split the `--admin` out into a separate array item

File: app/Http/Controllers/Admin/ProjectController.php [hunk 3]
```diff
@@ -51,12 +52,16 @@ public function index(
     ) {
         $projects = $this->repository->getAll();
 
+        // Getting all users
+        $users = User::where('is_admin', 0)->get();
```
Comment (REBELinBLUE) [inline app/Http/Controllers/Admin/ProjectController.php]: This should probably be changed a method on the user repository and inject the repository in the method

File: app/Http/Middleware/IsAdmin.php [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+     * Handle an incoming request.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @param  \Closure  $next
+     * @return mixed
+     */
+    public function handle($request, Closure $next)
+    {
+        // If user is authenticated ...
+        if (Auth::user()) {
```
Comment (REBELinBLUE) [inline app/Http/Middleware/IsAdmin.php]: Take a look at the Authenticate middle, I am injecting the AuthFactory as I really don't like Laravel's "facades"

File: app/Policies/ProjectPolicy.php [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+use REBELinBLUE\Deployer\Project;
+use Illuminate\Auth\Access\HandlesAuthorization;
+
+class ProjectPolicy
+{
+    use HandlesAuthorization;
+
+    public function before($user, $ability)
+    {
+        if ($user->is_admin === 1) {
```
Comment (REBELinBLUE) [inline app/Policies/ProjectPolicy.php]: Add `is_admin` to the `casts` array on the `User` entity to cast it to a boolean then you can just do `$user->is_admin`

File: app/Console/Commands/CreateUser.php (hunk 1 above)
Comment (axeloz) [reply app/Console/Commands/CreateUser.php]: Done

File: app/Console/Commands/InstallApp.php (hunk 2 above)
Comment (axeloz) [reply app/Console/Commands/InstallApp.php]: Done

File: app/Policies/ProjectPolicy.php (hunk 5 above)
Comment (axeloz) [reply app/Policies/ProjectPolicy.php]: Done

Comment (axeloz) [reply app/Policies/ProjectPolicy.php]: Oops, breaks my code.

Comment (axeloz) [reply app/Policies/ProjectPolicy.php]: fixed

File: app/Http/Middleware/IsAdmin.php (hunk 4 above)
Comment (axeloz) [reply app/Http/Middleware/IsAdmin.php]: I will check this in the morning

**Comment by REBELinBLUE** [comment]
In the servers view you can replace the autocomplete with the following

```$(`#${element} #${element}_name`).typeahead({
  autoSelect: false,
  source: (query, process) => $.ajax({
    type: 'GET',
    url: routes.serversAutocomplete,
    data: { query },
  }).done(response => process($.map(response.suggestions, dataItem => ({
    name: `${dataItem.name} (${dataItem.user}@${dataItem.ip_address})`,
    data: dataItem,
  })))),
  afterSelect: (suggestion) => {
    fields.forEach((field) => {
      $(`#${element}_${field}`).val(suggestion.data[field]);
    });

    $(`#${element}_deploy_code`).prop('checked', suggestion.data.deploy_code);
  },
});```

Then you can remove `devbridge-autocomplete` from the `package.json`, no point in having 2 different autocomplete packages

**Comment by REBELinBLUE** [comment]
Here is the autocomplete for the users. It still needs some work but should be enough to get you started

```javascript
/**
 * Autocomplete feature for project's members
 */
$('.members_autocomplete').tagsinput({
  allowDuplicates: false,
  freeInput: false,
  itemText: 'text',
  itemValue: 'value',
  typeahead: {
    name: 'users',
    source: (query) => {
      return UserCollection.filter((user) => {
        // Move this to a method on the collection instead
        return user.get('name').toLowerCase().startsWith(query.toLowerCase());
      }).map(user => ({
        id: user.get('id'),
        text: user.get('name'),
      }));
    },
    afterSelect: function() {
    	this.$element[0].value = '';
    }
  }
});
```

You obviously need to add `import UserCollection from '../../collections/Users';`

Whilst you making changes could you fix a mistake for me, in `src/Models/User.js` the class is called `Group`, can you change it to `User`

**Comment by axeloz** [comment]
Hello, I did change everything according to your needs. 
I'm trying to fix a PHP exception and it should be ok after this. 
Thanks for your help

Axel

**Comment by axeloz** [comment]
All good !

**Comment by REBELinBLUE** [comment]
So are you finished?

Tests fail

**Comment by axeloz** [comment]
Oops, errors yep. Because of my coding style. 
I'm fixing right now...

**Comment by REBELinBLUE** [comment]
There are tests which will need updating as well https://travis-ci.org/REBELinBLUE/deployer/jobs/395025198

**Comment by axeloz** [comment]
Yeah I've seen that. In progress but I have regular work to do right now so pending ... :)

**Comment by REBELinBLUE** [comment]
Ha no worries, just wasn't sure if you had seen :)

**Comment by axeloz** [comment]
Hello @REBELinBLUE I have some issues fixing two errors during Unit testing (one of them is recurrent):

```
1) REBELinBLUE\Deployer\Tests\Unit\Repositories\EloquentProjectRepositoryTest::testGetAll
Mockery\Exception\BadMethodCallException: Static method Mockery_16_REBELinBLUE_Deployer_Project::with() does not exist on this mock object

/vagrant/www/deployer/app/Repositories/EloquentProjectRepository.php:36
/vagrant/www/deployer/tests/Unit/Repositories/EloquentProjectRepositoryTest.php:51

Caused by
Mockery\Exception\BadMethodCallException: Received Mockery_16_REBELinBLUE_Deployer_Project::with(), but no expectations were specified

/vagrant/www/deployer/app/Repositories/EloquentProjectRepository.php:36
/vagrant/www/deployer/tests/Unit/Repositories/EloquentProjectRepositoryTest.php:51

2) REBELinBLUE\Deployer\Tests\Unit\Repositories\EloquentProjectRepositoryTest::testUpdateByIdRemovesBlankPrivateKey
Mockery\Exception\BadMethodCallException: Received Mockery_16_REBELinBLUE_Deployer_Project::users(), but no expectations were specified

/vagrant/www/deployer/app/Repositories/EloquentProjectRepository.php:136
/vagrant/www/deployer/app/Repositories/EloquentProjectRepository.php:102
/vagrant/www/deployer/tests/Unit/Repositories/EloquentProjectRepositoryTest.php:143
``` 

About the second one, I guess this is because of the users() relation I added in the projects model. 
Any idea? 
Thanks

**Comment by REBELinBLUE** [comment]
You've added `->with('users')` to the `getAll()` method in the `EloquentProjectRepository`.

If you look at `testGetAll` in the `EloquentProjectRepositoryTest` the model is a mock

```
        $model  = m::mock(Project::class);
        $model->shouldReceive('orderBy')->once()->with('name')->andReturn($expected);
```

so you need to include `with` in the expectation

```
        $model  = m::mock(Project::class);
        $model->shouldReceive('with')->with('user')->andReturnSelf();
        $model->shouldReceive('orderBy')->with('name')->andReturn($expected);
```

looks like the errors are all basically the same sort of thing

**Comment by axeloz** [comment]
Sorry, I tried that but it did not fix, I'm still stuck. I don't get it...

**Comment by REBELinBLUE** [comment]
Here is a patch file to fix the unit tests


[mypatch.txt](https://github.com/REBELinBLUE/deployer/files/2164234/mypatch.txt)

**Comment by axeloz** [comment]
@REBELinBLUE thanks man, tests should run OK now....

**Comment by axeloz** [comment]
Thanks; new push to fix these errors and warnings. Yep, it seems to conflict. I'll try to figure out why but not sure how to fix yet.

**Comment by REBELinBLUE** [comment]
my fault, it is because I changed the javascript files. Easiest way to resolve it is to make sure your copy of master is up to date with my master https://help.github.com/articles/syncing-a-fork/

Then on your branch do `git rebase master`, you'll get a merge conflict but it doesn't matter in this case, just do `git add -A`, the `git rebase --continue`, then regenerate the JS and commit it again

**Comment by axeloz** [comment]
Cool it passed. It's been harder to write the unit testing than adding this feature ...

**Comment by REBELinBLUE** [comment]
Can you open the PR for the frontend code as well so I can merge this?

**Comment by axeloz** [comment]
Thanks, just added a merge request.

**Comment by axeloz** [comment]
Hello @REBELinBLUE is everything OK ? Thanks

**Comment by REBELinBLUE** [comment]
Yeah sorry just been really busy and haven't had a chance to go through everything yet, will try to before the end of the week

**Comment by REBELinBLUE** [comment]
Sorry for the delay, you know how life gets in the way sometimes, I've set aside time tomorrow evening to go over this

**Comment by axeloz** [comment]
Hello @REBELinBLUE, I know exactly what you mean so don't worry. Let me know
Axel

**Comment by REBELinBLUE** [comment]
Mainly looks good, only one small thing I can see, can you update the migration so that all the existing users have is_admin set to true so that we don't break existing installs?

**Pull request merged**